### ISSUE TITLE
[Table] cells can truncate content, show tooltip if truncated

### ISSUE THREAD

**Issue by davidicus** [IBM/carbon-addons-iot-react#1051]
Copied from internal repo from @KarenGski 
> **Is this an extension to an existing component, pattern, or a brand new asset?**
> 
> **What is the rationale for this contribution.**
> In data tables, we need to allow the content to truncate so that the rows don't become very very tall if the content is longer that the set column widgth. If the content is trunctated, the user should be able to click on it to see the full, untruncated content.
> 
> **Detail the intended behavior**
> Include mockups of any fidelity, the more spec the better as this will enable higher quality feedback and evaluation.
> 
> Include for accessibility
> 
> * Examples of the elements in various states
> * Expected keyboard behavior
> * Alt text
> 
> **Inspiration from other products**
> (Optional - remove this is you don't have any)

**Comment by davidicus** [comment]
Copied from internal repo

@JordanWSmith15 
> This exists in the PAL, but is not implemented into the data table in the iot repo. Third to last screen shot at https://pages.github.ibm.com/Watson-IoT/IoT-Design-Site/components/datatable/basictable
> ![Screen Shot 2020-03-26 at 10 37 51 AM](https://media.github.ibm.com/user/17018/files/e5652f00-6f4d-11ea-855e-b4598155f4b6)

@tay1orjones 
> While this is not a directly implemented feature, it is currently supported via the [`renderDataFunction`](https://carbon-addons-iot-react.com/?path=/story/watson-iot-table--with-custom-cell-renderer), which allows custom nodes/components to be placed within table cells.
> 
> There have been some recent bug fixes to ensure [this works in TableCards](https://github.com/IBM/carbon-addons-iot-react/pull/990), as well as [when sticky headers are enabled](https://github.com/IBM/carbon-addons-iot-react/pull/906).

@JordanWSmith15 
> @kgoscimi @taylorjones @lukefirth I think that, based on our slack conversation, that this story is being used to set the default behavior of the table back to truncate, rather than wrap?

**Comment by bjornalm** [comment]
@lukefirth should the column header texts also truncate or should they continue to wrap?

**Comment by lukefirth** [comment]
Great question. Personally, I think for consistency in expectation we should go with truncate.

For posterity if we come back to look at this issue in future - part of my reasoning process:

- If we let it wrap, it would be less consistent with the overall table behavior, and we open ourselves up to potentially a user seeing a wrapped header, deciding they prefer wrapping to truncation and then being annoyed we don't do that for everything.

- If we go with truncation, the header behaves like the rest of the table (great), and we do not advertise an alternative behavior for a user to see, possibly think they prefer and then be annoyed that they can't have it 😄 


Truncation wins, until / unless more data comes up in the future that outweighs the logical balance outlined above.

**Comment by bjornalm** [comment]
@lukefirth To clarify, the cell text should only be truncated if the column has an explicit width (i.e. fixed column width tables or tables with column resize)?

Today, most stories show tables with column widths determined by the content. Those columns will grow horizontally if possible and wrap content if needed. This is handled by the browser.  It is possible to add truncation to these tables a well but it will require more

**Comment by bjornalm** [comment]
@lukefirth To clarify, should the cell text only be truncated if the column has an explicit width (i.e. fixed column width tables or tables with column resize)? If so, that is good. If not things gets more complicated.

Today, most stories show tables with column widths determined by the content. Those columns will grow horizontally if possible and wrap content if needed. This is well handled by the browser and standard css. It is possible to also add truncation in these scenarios but it will be more complex, either by more javascript that probably impacts the performance, or by "CSS hacks" like in the example I created here: http://jsfiddle.net/5owf1zck/5/

**Comment by davidicus** [comment]
@JordanWSmith15 @KarenGski we are good with truncation and resize feature being mutually exclusive?

**Comment by davidicus** [comment]
> Great question. Personally, I think for consistency in expectation we should go with truncate.
> 
> For posterity if we come back to look at this issue in future - part of my reasoning process:
> 
> * If we let it wrap, it would be less consistent with the overall table behavior, and we open ourselves up to potentially a user seeing a wrapped header, deciding they prefer wrapping to truncation and then being annoyed we don't do that for everything.
> * If we go with truncation, the header behaves like the rest of the table (great), and we do not advertise an alternative behavior for a user to see, possibly think they prefer and then be annoyed that they can't have it 😄
> 
> Truncation wins, until / unless more data comes up in the future that outweighs the logical balance outlined above.

@lukefirth it was my understanding that this would be a feature you can opt in and out of via props. So a user can choose whatever the prefer. Def true that the wrapping of cells and not headers is inconsistent.

**Comment by lukefirth** [comment]
Gosh, there really are a lot of details hidden *just* beneath the surface with all of this stuff ha.

I'll outline some expectation scenarios and see how we fair. Please give me back and forth on this. I think again if the table has a horizontal scroll behavior or not may be a deciding factor as to how we handle the columns:

In all cases:
- If the end user sets a specific width of the column, it should trigger truncate.

In horizontally scrolling cases:
- By default, columns should grow based on content (a max-width could have some value here however).
- As they grow, the table gets wider but this is handled by the horizontal scroll.


In cases where horizontal scrolling is not enabled (so the table is contained to a certain width).
- Columns should be taking up a default % of the available width, determined by how many columns there are (unless resized, which would allow them to take on more or less).
- If their content extends beyond their allocated width based on the above %, then they should truncate.

----

In other words:

Truncation should happen:
- Whenever a user manually sets the size to smaller than the content of the column.
- In a fixed-width (no horizontal scroll) table, where the content of a column exceeds the default width allocated to it based on the available space (X% of total table width, based on column numbers of the
